File: containers/bench.py
# /// script
# requires-python = ">=3.13"
# dependencies = [
#   "stringzilla",
#   "rbloom",
#   "cuckoofilter",
#   "xxhash",
# ]
# ///
"""
Multi-way word hashing and probabilistic membership benchmarks, mirroring containers/bench.rs.

Layer 1 produces `k` hashes per word for `k` in {2, 4, 8, 16}:
- StringZilla `hash_multiseed` emits all `k` hashes from one input preparation.
- StringZilla `hash` called once per seed, isolating what the multi-seed path amortizes.
- One 128-bit `xxh3_128` split into `(h1, h2)` with `g_i = h1 + i*h2` (the double-hashing baseline).

Layer 2 builds a Bloom filter (rbloom, default vs StringZilla-fed) and a cuckoo filter (cuckoofilter),
then queries a held-out set to measure the false-positive rate.

Environment variables:
- STRINGWARS_DATASET: Path to input dataset file
- STRINGWARS_TOKENS: Tokenization mode ('lines', 'words', 'file')

Examples:
  uv run containers/bench.py --dataset xlsum.csv --tokens words
  uv run containers/bench.py --dataset data.txt --tokens words -k "multiseed"
"""

import argparse
import re
import sys
from array import array
from collections.abc import Callable

import stringzilla as sz
import xxhash
from cuckoofilter import CuckooFilter
from rbloom import Bloom

from utils import (
    add_common_args,
    load_dataset,
    now_nanoseconds,
    paced_items,
    report_stats,
    resolve_tokens,
    should_run,
    tokenize_dataset,
)

# Sixteen fixed odd seeds shared by every multi-hash variant, matching containers/bench.rs.
SEEDS = [
    0x9E3779B97F4A7C15,
    0xC2B2AE3D27D4EB4F,
    0x165667B19E3779F9,
    0xD1B54A32D192ED03,
    0xA0761D6478BD642F,
    0xE7037ED1A0B428DB,
    0x8EBC6AF09C88C6E3,
    0x589965CC75374CC3,
    0x1D8E4E27C47D124F,
    0xEB44ACCAB455D165,
    0x2545F4914F6CDD1D,
    0xFF51AFD7ED558CCD,
    0xC4CEB9FE1A85EC53,
    0xBF58476D1CE4E5B9,
    0x94175CC1BAB35C97,
    0x4CF5AD432745937F,
]

MASK_64 = (1 << 64) - 1
TARGET_FALSE_POSITIVE_RATE = 0.01


def stringzilla_hash_128(data: bytes) -> int:
    """Two seeded StringZilla hashes combined into the signed 128-bit integer rbloom expects."""
    value = (sz.hash(data, SEEDS[0]) << 64) | sz.hash(data, SEEDS[1])
    return value - (1 << 128) if value >= (1 << 127) else value


def log_system_info():
    print(f"- Python: {sys.version.split()[0]}, {sys.platform}")
    print(f"- StringZilla: {sz.__version__} with {sz.__capabilities_str__}")
    print(f"- xxHash: {xxhash.VERSION}")
    print()


def bench_multihash(name: str, tokens: list[bytes], produce: Callable[[bytes], object], k: int, time_limit: float):
    """Time one multi-hash variant, calling `produce` once per word to emit `k` hashes."""
    start = now_nanoseconds()
    deadline = start + int(time_limit * 1e9)
    count = 0
    total_bytes = 0
    for token in paced_items(tokens, deadline):
        produce(token)
        count += 1
        total_bytes += len(token)
    seconds = (now_nanoseconds() - start) / 1e9
    report_stats(name, "hashes", seconds, count * k, total_bytes)


def bench_build(name: str, count: int, total_bytes: int, build: Callable[[], object], time_limit: float):
    """Time filter construction, rebuilding from every key on each pass."""
    start = now_nanoseconds()
    deadline = start + int(time_limit * 1e9)
    passes = 0
    while True:
        build()
        passes += 1
        if now_nanoseconds() >= deadline:
            break
    seconds = (now_nanoseconds() - start) / 1e9
    report_stats(name, "hashes", seconds, passes * count, passes * total_bytes)


def bench_query(name: str, probes: list[bytes], query: Callable[[bytes], bool], time_limit: float):
    """Time membership queries, cycling one probe per call."""
    start = now_nanoseconds()
    deadline = start + int(time_limit * 1e9)
    count = 0
    total_bytes = 0
    for token in paced_items(probes, deadline):
        query(token)
        count += 1
        total_bytes += len(token)
    seconds = (now_nanoseconds() - start) / 1e9
    report_stats(name, "hashes", seconds, count, total_bytes)


def report_quality(
    label: str, num_bits: int, inserted_count: int, absent: list[bytes], contains: Callable[[bytes], bool]
):
    """Report the measured false-positive rate over the held-out absent words plus bits-per-key."""
    false_positives = sum(1 for token in absent if contains(token))
    rate = false_positives / len(absent) * 100 if absent else 0.0
    bits = f"{num_bits / max(inserted_count, 1):5.2f} bits/key" if num_bits else "    n/a    "
    print(f"    {label:<38} {bits}, measured FPR {rate:.3f}%")


def run_multihash(tokens: list[bytes], filter_pattern: re.Pattern | None, time_limit: float):
    for k in (2, 4, 8, 16):
        print(f"# multihash (k={k})")
        seeds = array("Q", SEEDS[:k])
        out = array("Q", [0] * k)

        if should_run("multihash/stringzilla.hash_multiseed", filter_pattern):
            bench_multihash(
                "multihash/stringzilla.hash_multiseed",
                tokens,
                lambda token, seeds=seeds, out=out: sz.hash_multiseed(token, seeds, out),
                k,
                time_limit,
            )
        if should_run("multihash/stringzilla.hash", filter_pattern):
            bench_multihash(
                "multihash/stringzilla.hash",
                tokens,
                lambda token, seeds=seeds: [sz.hash(token, seed) for seed in seeds],
                k,
                time_limit,
            )
        if should_run("multihash/xxhash.xxh3_128", filter_pattern):

            def double_hash(token, k=k):
                wide = xxhash.xxh3_128(token).intdigest()
                first, second = wide & MASK_64, wide >> 64
                return [(first + index * second) & MASK_64 for index in range(k)]

            bench_multihash("multihash/xxhash.xxh3_128", tokens, double_hash, k, time_limit)


def run_filters(unique: list[bytes], filter_pattern: re.Pattern | None, time_limit: float):
    inserted_count = min(len(unique) * 8 // 10, 1_000_000) or 1
    inserted = unique[:inserted_count]
    absent = unique[inserted_count:]
    inserted_bytes = sum(len(token) for token in inserted)
    print(f"# filters ({inserted_count} inserted, {len(absent)} held-out absent)")

    # Bloom filter (rbloom): default Python hash versus a StringZilla-fed hash function.
    bloom = Bloom(inserted_count, TARGET_FALSE_POSITIVE_RATE)
    for token in inserted:
        bloom.add(token)
    report_quality("bloom/rbloom<siphash>", bloom.size_in_bits, inserted_count, absent, lambda t: t in bloom)
    if should_run("bloom/rbloom.insert<siphash>", filter_pattern):

        def build_bloom():
            filter_ = Bloom(inserted_count, TARGET_FALSE_POSITIVE_RATE)
            for token in inserted:
                filter_.add(token)

        bench_build("bloom/rbloom.insert<siphash>", inserted_count, inserted_bytes, build_bloom, time_limit)
    if should_run("bloom/rbloom.contains<siphash>", filter_pattern):
        bench_query("bloom/rbloom.contains<siphash>", inserted, lambda t: t in bloom, time_limit)

    bloom_sz = Bloom(inserted_count, TARGET_FALSE_POSITIVE_RATE, stringzilla_hash_128)
    for token in inserted:
        bloom_sz.add(token)
    report_quality("bloom/rbloom<stringzilla>", bloom_sz.size_in_bits, inserted_count, absent, lambda t: t in bloom_sz)
    if should_run("bloom/rbloom.insert<stringzilla>", filter_pattern):

        def build_bloom_sz():
            filter_ = Bloom(inserted_count, TARGET_FALSE_POSITIVE_RATE, stringzilla_hash_128)
            for token in inserted:
                filter_.add(token)

        bench_build("bloom/rbloom.insert<stringzilla>", inserted_count, inserted_bytes, build_bloom_sz, time_limit)
    if should_run("bloom/rbloom.contains<stringzilla>", filter_pattern):
        bench_query("bloom/rbloom.contains<stringzilla>", inserted, lambda t: t in bloom_sz, time_limit)

    # Cuckoo filter (cuckoofilter): pure-Python, hashes each key internally.
    cuckoo = CuckooFilter(capacity=inserted_count, fingerprint_size=2)
    for token in inserted:
        cuckoo.insert(token)
    report_quality("cuckoo/cuckoofilter", 0, inserted_count, absent, lambda t: cuckoo.contains(t))
    if should_run("cuckoo/cuckoofilter.insert", filter_pattern):

        def build_cuckoo():
            filter_ = CuckooFilter(capacity=inserted_count, fingerprint_size=2)
            for token in inserted:
                filter_.insert(token)

        bench_build("cuckoo/cuckoofilter.insert", inserted_count, inserted_bytes, build_cuckoo, time_limit)
    if should_run("cuckoo/cuckoofilter.contains", filter_pattern):
        bench_query("cuckoo/cuckoofilter.contains", inserted, lambda t: cuckoo.contains(t), time_limit)


def main():
    parser = argparse.ArgumentParser(
        description="Benchmark multi-way word hashing and probabilistic membership structures",
        formatter_class=argparse.RawDescriptionHelpFormatter,
    )
    add_common_args(parser)
    args = parser.parse_args()

    filter_pattern = None
    if args.filter:
        try:
            filter_pattern = re.compile(args.filter)
        except re.error as error:
            parser.error(f"Invalid regex for --filter: {error}")

    dataset = load_dataset(args.dataset, as_bytes=True, size_limit=args.dataset_limit)
    tokens = tokenize_dataset(dataset, resolve_tokens(args.tokens, "words"))
    if not tokens:
        print("No tokens found in dataset")
        return 1

    unique = list(dict.fromkeys(tokens))
    total_bytes = sum(len(token) for token in tokens)
    print(f"Dataset: {len(tokens):,} tokens, {total_bytes:,} bytes, {len(unique):,} unique")
    log_system_info()

    run_multihash(tokens, filter_pattern, args.time_limit)
    run_filters(unique, filter_pattern, args.time_limit)
    return 0


if __name__ == "__main__":
    exit(main())

File: containers/bench.rs
#![doc = r#"
# StringWars: Multi-Way Word Hashing & Probabilistic Membership Benchmarks

Probabilistic membership structures — Bloom, cuckoo, and XOR/binary-fuse filters — need several independent hashes of
the same short key. This file benchmarks that operation in two layers.

**Layer 1 — multi-hash generation** produces `k` hashes per word for `k` in `{2, 4, 8, 16}`:

- StringZilla `hash_multiseed` normalizes the input into AES blocks once, then replays cheap per-seed rounds, emitting
  all `k` hashes in a single pass.
- StringZilla `hash` called once per seed, isolating what the multi-seed path amortizes.
- One 128-bit `xxh3_128` split into `(h1, h2)` with `g_i = h1 + i*h2` (Kirsch–Mitzenmacher) — the double-hashing scheme
  production Bloom and cuckoo filters actually use.

**Layer 2 — probabilistic membership** builds each filter from the unique words and queries a held-out set to measure
the false-positive rate, comparing a StringZilla-fed variant against the practical default while holding the filter
structure fixed.

## Usage

```sh
STRINGWARS_DATASET=xlsum.csv \
    STRINGWARS_TOKENS=words \
    cargo bench --features bench_containers --bench bench_containers
```
"#]

use std::collections::hash_map::DefaultHasher;
use std::collections::HashSet;
use std::hint::black_box;

use stringtape::{BytesCowsAuto, BytesTape};

use cuckoofilter::CuckooFilter;
use fastbloom::BloomFilter;
use stringzilla::sz;
use xorf::{BinaryFuse8, Filter};
use xxhash_rust::xxh3::{xxh3_128, xxh3_64};

#[path = "../utils.rs"]
mod utils;
use utils::{
    install_panic_hook, load_dataset_with_default_mode, log_stringzilla_metadata,
    measure_throughput, should_run, BenchBudget, ReportAs, ResultExt, WorkUnits,
};

/// Sixteen fixed odd seeds, enough for the largest `k`, shared across every multi-hash variant so
/// each produces the same family of hashes.
const SEEDS: [u64; 16] = [
    0x9E37_79B9_7F4A_7C15,
    0xC2B2_AE3D_27D4_EB4F,
    0x1656_67B1_9E37_79F9,
    0xD1B5_4A32_D192_ED03,
    0xA076_1D64_78BD_642F,
    0xE703_7ED1_A0B4_28DB,
    0x8EBC_6AF0_9C88_C6E3,
    0x5899_65CC_7537_4CC3,
    0x1D8E_4E27_C47D_124F,
    0xEB44_ACCA_B455_D165,
    0x2545_F491_4F6C_DD1D,
    0xFF51_AFD7_ED55_8CCD,
    0xC4CE_B9FE_1A85_EC53,
    0xBF58_476D_1CE4_E5B9,
    0x9417_5CC1_BAB3_5C97,
    0x4CF5_AD43_2745_937F,
];

/// The target false-positive rate configured for the Bloom and cuckoo filters.
const TARGET_FALSE_POSITIVE_RATE: f64 = 0.01;

/// Times one multi-hash variant, calling `fill` once per word to produce `k` hashes.
fn measure_multihash<Fill: FnMut(&[u8])>(
    name: &str,
    budget: &BenchBudget,
    k: usize,
    slices: &[&[u8]],
    mut fill: Fill,
) {
    if !should_run(name) {
        return;
    }
    let mut cursor = 0usize;
    measure_throughput(name, ReportAs::Hashes, budget, || {
        let token = slices[cursor % slices.len()];
        cursor += 1;
        fill(token);
        WorkUnits::new(k as u64, token.len() as u64)
    });
}

/// Times filter construction, rebuilding the whole filter from `count` keys on every pass.
fn measure_build<Build: FnMut()>(
    name: &str,
    budget: &BenchBudget,
    count: usize,
    bytes: u64,
    mut build: Build,
) {
    if !should_run(name) {
        return;
    }
    measure_throughput(name, ReportAs::Hashes, budget, || {
        build();
        WorkUnits::new(count as u64, bytes)
    });
}

/// Times membership queries, cycling one probe per call.
fn measure_query<Query: FnMut(&[u8]) -> bool>(
    name: &str,
    budget: &BenchBudget,
    probes: &[&[u8]],
    mut query: Query,
) {
    if !should_run(name) {
        return;
    }
    let mut cursor = 0usize;
    measure_throughput(name, ReportAs::Hashes, budget, || {
        let token = probes[cursor % probes.len()];
        cursor += 1;
        black_box(query(token));
        WorkUnits::new(1, token.len() as u64)
    });
}

/// Reports the measured false-positive rate over the held-out absent words plus bits-per-key.
fn report_quality<Contains: FnMut(&[u8]) -> bool>(
    label: &str,
    num_bits: usize,
    inserted_count: usize,
    absent: &[&[u8]],
    mut contains: Contains,
) {
    let false_positives = absent.iter().filter(|token| contains(token)).count();
    let rate = if absent.is_empty() {
        0.0
    } else {
        false_positives as f64 / absent.len() as f64 * 100.0
    };
    let bits_per_key = num_bits as f64 / inserted_count.max(1) as f64;
    println!(
        "    {:<38} {:5.2} bits/key, measured FPR {:.3}%",
        label, bits_per_key, rate
    );
}

/// Layer 1: produce `k` hashes per word, reporting hashes/s.
fn bench_multihash(budget: &BenchBudget, k: usize, slices: &[&[u8]]) {
    println!("# multihash (k={})", k);
    let mut hashes = [0u64; 16];

    measure_multihash(
        "multihash/stringzilla::hash_multiseed",
        budget,
        k,
        slices,
        |token| {
            sz::hash_multiseed_into(token, &SEEDS[..k], &mut hashes[..k]);
            black_box(&hashes[..k]);
        },
    );

    measure_multihash("multihash/stringzilla::hash", budget, k, slices, |token| {
        for (slot, seed) in hashes[..k].iter_mut().zip(&SEEDS[..k]) {
            *slot = sz::hash_with_seed(token, *seed);
        }
        black_box(&hashes[..k]);
    });

    measure_multihash("multihash/xxh3::xxh3_128", budget, k, slices, |token| {
        let wide = xxh3_128(token);
        let (first, second) = (wide as u64, (wide >> 64) as u64);
        for (index, slot) in hashes[..k].iter_mut().enumerate() {
            *slot = first.wrapping_add((index as u64).wrapping_mul(second));
        }
        black_box(&hashes[..k]);
    });
}

/// Bloom filter (fastbloom): SipHash default versus a single `sz::hash` fed through `insert_hash`.
fn bench_bloom(budget: &BenchBudget, inserted: &[&[u8]], absent: &[&[u8]], bytes: u64) {
    let count = inserted.len();

    let mut bloom = BloomFilter::with_false_pos(TARGET_FALSE_POSITIVE_RATE).expected_items(count);
    for token in inserted {
        bloom.insert(token);
    }
    report_quality(
        "bloom/fastbloom<siphash>",
        bloom.num_bits(),
        count,
        absent,
        |token| bloom.contains(token),
    );
    measure_build(
        "bloom/fastbloom::insert<siphash>",
        budget,
        count,
        bytes,
        || {
            let mut filter =
                BloomFilter::with_false_pos(TARGET_FALSE_POSITIVE_RATE).expected_items(count);
            for token in inserted {
                filter.insert(token);
            }
            black_box(&filter);
        },
    );
    measure_query(
        "bloom/fastbloom::contains<siphash>",
        budget,
        inserted,
        |token| bloom.contains(token),
    );

    let mut bloom_sz =
        BloomFilter::with_false_pos(TARGET_FALSE_POSITIVE_RATE).expected_items(count);
    for token in inserted {
        bloom_sz.insert_hash(sz::hash(token));
    }
    report_quality(
        "bloom/fastbloom<stringzilla>",
        bloom_sz.num_bits(),
        count,
        absent,
        |token| bloom_sz.contains_hash(sz::hash(token)),
    );
    measure_build(
        "bloom/fastbloom::insert<stringzilla>",
        budget,
        count,
        bytes,
        || {
            let mut filter =
                BloomFilter::with_false_pos(TARGET_FALSE_POSITIVE_RATE).expected_items(count);
            for token in inserted {
                filter.insert_hash(sz::hash(token));
            }
            black_box(&filter);
        },
    );
    measure_query(
        "bloom/fastbloom::contains<stringzilla>",
        budget,
        inserted,
        |token| bloom_sz.contains_hash(sz::hash(token)),
    );
}

/// Cuckoo filter (cuckoofilter): SipHash default. `cuckoofilter` hashes the key twice per op
/// (once for the fingerprint, once for the alternate bucket) through the `Hasher` trait and exposes
/// no precomputed-hash entry point, so there is no StringZilla-fed variant to compare here.
fn bench_cuckoo(budget: &BenchBudget, inserted: &[&[u8]], absent: &[&[u8]], bytes: u64) {
    let count = inserted.len();

    let mut cuckoo = CuckooFilter::<DefaultHasher>::with_capacity(count);
    for token in inserted {
        let _ = cuckoo.add(token);
    }
    report_quality(
        "cuckoo/cuckoofilter",
        cuckoo.memory_usage() * 8,
        count,
        absent,
        |token| cuckoo.contains(token),
    );
    measure_build("cuckoo/cuckoofilter::insert", budget, count, bytes, || {
        let mut filter = CuckooFilter::<DefaultHasher>::with_capacity(count);
        for token in inserted {
            let _ = filter.add(token);
        }
        black_box(&filter);
    });
    measure_query("cuckoo/cuckoofilter::contains", budget, inserted, |token| {
        cuckoo.contains(token)
    });
}

/// Collects the keys into a sorted, deduplicated `u64` array — the shape `xorf` consumes.
fn hashed_keys<Hash: Fn(&[u8]) -> u64>(inserted: &[&[u8]], hash: Hash) -> Vec<u64> {
    let mut keys: Vec<u64> = inserted.iter().map(|token| hash(token)).collect();
    keys.sort_unstable();
    keys.dedup();
    keys
}

/// Binary-fuse filter (xorf): a static filter built from pre-hashed keys, so the only variable is the
/// hash that produced them. Its ~0.4% false-positive rate is fixed by the 8-bit fingerprints.
fn bench_xorf(budget: &BenchBudget, inserted: &[&[u8]], absent: &[&[u8]], bytes: u64) {
    let count = inserted.len();

    let fuse_sz = BinaryFuse8::try_from(&hashed_keys(inserted, |token| sz::hash(token)))
        .expect("binary-fuse build from StringZilla keys");
    let fuse_xxh = BinaryFuse8::try_from(&hashed_keys(inserted, xxh3_64))
        .expect("binary-fuse build from xxh3 keys");
    report_quality(
        "xor/xorf::BinaryFuse8<stringzilla>",
        fuse_sz.len() * 8,
        count,
        absent,
        |token| fuse_sz.contains(&sz::hash(token)),
    );
    report_quality(
        "xor/xorf::BinaryFuse8<xxh3>",
        fuse_xxh.len() * 8,
        count,
        absent,
        |token| fuse_xxh.contains(&xxh3_64(token)),
    );

    measure_build(
        "xor/xorf::BinaryFuse8::build<stringzilla>",
        budget,
        count,
        bytes,
        || {
            black_box(
                BinaryFuse8::try_from(&hashed_keys(inserted, |token| sz::hash(token))).unwrap(),
            );
        },
    );
    measure_query(
        "xor/xorf::BinaryFuse8::contains<stringzilla>",
        budget,
        inserted,
        |token| fuse_sz.contains(&sz::hash(token)),
    );
    measure_build(
        "xor/xorf::BinaryFuse8::build<xxh3>",
        budget,
        count,
        bytes,
        || {
            black_box(
                BinaryFuse8::try_from(&hashed_keys(inserted, xxh3_64)).unwrap(),
            );
        },
    );
    measure_query(
        "xor/xorf::BinaryFuse8::contains<xxh3>",
        budget,
        inserted,
        |token| fuse_xxh.contains(&xxh3_64(token)),
    );
}

/// Layer 2: build and query each filter, holding out 20% of the unique words as absent probes.
fn bench_filters(budget: &BenchBudget, unique: &[&[u8]]) {
    let inserted_count = (unique.len() * 8 / 10).clamp(1, 1_000_000);
    let inserted = &unique[..inserted_count];
    let absent = &unique[inserted_count..];
    let inserted_bytes: u64 = inserted.iter().map(|token| token.len() as u64).sum();

    println!(
        "# filters ({} inserted, {} held-out absent)",
        inserted_count,
        absent.len()
    );
    bench_bloom(budget, inserted, absent, inserted_bytes);
    bench_cuckoo(budget, inserted, absent, inserted_bytes);
    bench_xorf(budget, inserted, absent, inserted_bytes);
}

/// Asserts the amortized multi-seed path emits exactly the hashes per-seed hashing would.
fn verify_multiseed_matches_naive() {
    let probe: &[u8] = b"multiseed-correctness-probe";
    let mut amortized = [0u64; 8];
    sz::hash_multiseed_into(probe, &SEEDS[..8], &mut amortized);
    for index in 0..8 {
        assert_eq!(
            amortized[index],
            sz::hash_with_seed(probe, SEEDS[index]),
            "hash_multiseed disagrees with hash_with_seed at seed index {}",
            index
        );
    }
    println!("- multiseed == naive per-seed hashing: OK");
}

fn main() {
    install_panic_hook();
    log_stringzilla_metadata();
    verify_multiseed_matches_naive();

    let tokens: BytesCowsAuto = load_dataset_with_default_mode("words").unwrap_nice();

    let mut tape = BytesTape::<u64>::new();
    tape.extend(tokens.iter())
        .expect("Failed to build BytesTape");
    let view = tape.view();
    let slices: Vec<&[u8]> = (&view).into_iter().collect();

    let unique: Vec<&[u8]> = {
        let set: HashSet<&[u8]> = slices.iter().copied().collect();
        set.into_iter().collect()
    };
    println!("- {} tokens, {} unique\n", slices.len(), unique.len());

    let budget = BenchBudget::from_env(2.0, 10.0);

    for k in [2usize, 4, 8, 16] {
        bench_multihash(&budget, k, &slices);
    }
    bench_filters(&budget, &unique);
}
